### PR TITLE
WebDAV backend (WIP)

### PR DESCRIPTION
**Pull request by jlbprof**

### CODE REVIEW

File: src/lib/adapters/WebDav.js [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+	}
+
+    createXBEL (myStructure) {
+        
+        let output = `<?xml version="1.0" encoding="ISO-8859-1"?>
+<!DOCTYPE xbel PUBLIC "+//IDN python.org//DTD XML Bookmark Exchange Language 1.0//EN//XML" "http://www.python.org/topics/xml/dtds/xbel-1.0.dtd">
+<xbel version="1.0">
+`;
+
+        output += '<!--- highestID :' + this.highestID + `: for Floccus bookmark sync browser extension -->
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: Nice idea! :)

File: src/lib/adapters/WebDav.js [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+		output += `
+</xbel>`;
+
+		return output;
+    }
+
+    /* private routine */
+    _addBookmark (myStructure, bm)
+    {
+        let myArray = bm.path.split ("/");
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: This won't work for all folder names, sadly. If they contain slashes, those are escaped with a backslash. Maybe we can find a better solution than to pass the path as a string. ([this](https://github.com/marcelklehr/floccus/blob/master/src/lib/Tree.js#L198) is how I parse those paths, theh)

File: src/lib/adapters/WebDav.js [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+        console.log ("WebDav: Uploading JSON file to server");
+        this.bookmarksAsJSON = this.getBookmarksAsJSON ();
+
+        let fullUrl = this.server.bookmark_file;
+        fullUrl = this.server.url + fullUrl;
+        console.log ("fullURL :" + fullUrl + ":");
+        let xbel = this.convertToStructure ();
+        this.uploadFile (fullUrl, 'application/xml', xbel);
+
+		this.freeLock ();
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: the above two lines probably each need an `await` ;)

File: src/lib/adapters/WebDav.js [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+            })
+        });
+
+        let folderList = this._getElementsByNodeName (xbelObj.childNodes, 'folder', 1 /* element type */);
+
+        console.log ("_parseFolder: 003");
+        console.log (folderList);
+
+        folderList.forEach ((folder) => {
+            let newpath = path + "/" + folder.firstElementChild.innerHTML;
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: slashes need escaping here, too :)

File: src/lib/adapters/WebDav.js [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+            throw new Error('Failed to fetch bookmark');
+        }
+        let bookmark = new Bookmark(bm.id, null, bm.url, bm.title, bm.path);
+        return bookmark;
+    }
+
+    async createBookmark (bm) {
+        console.log('Create bookmark: 001 ', bm, this.server)
+
+        // if highestID is zero than we have a new situation
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: :+1: for handling this! :)

File: src/lib/adapters/WebDav.js [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+
+        console.log('Create bookmark: OUT ', bm, this.server)
+
+        return bm;
+    }
+
+    async updateBookmark (remoteId, newBm) {
+        console.log('Update bookmark', newBM, remoteId, this.server)
+
+        let bm = await this.getBookmark(remoteId, false)
```
Comment (marcelklehr) [inline src/lib/adapters/WebDav.js]: this is probably unnecessary, but that's being pickish :D

File: src/lib/Account.js [hunk 7]
```diff
@@ -160,7 +163,11 @@ export default class Account {
 
       await this.tree.removeOrphanedFolders()
 
+      if ('syncComplete' in this.server)
+        await this.server.syncComplete ();
+
       await this.setData({...this.getData(), error: null, syncing: false, lastSync: Date.now()})
+
       console.log('Successfully ended sync process for account ' + this.getLabel())
     } catch (e) {
```
Comment (marcelklehr) [inline src/lib/Account.js]: In case of an error, the lock file is not deleted.

**Comment by marcelklehr** [comment]
I cannot get it to work. Probably because I don't have an existing filled xbel file on my server?

**Comment by jlbprof** [comment]
are you referring to the "away ths.getBookmark"?

**Comment by jlbprof** [comment]
Man I cannot type

await this.getBookmark

**Comment by jlbprof** [comment]
Can you explain your regex?  I am confused because you first reverse it.
 What exactly is it searching for?

Thanx

Julian

**Comment by jlbprof** [comment]
The slashes are they how the actual bookmark path's are?   It would seem
odd to me that you could put a slash in a path value like that.

**Comment by marcelklehr** [comment]
> are you referring to the "away ths.getBookmark"?
yup :)

> Can you explain your regex?  I am confused because you first reverse it.

heh, sure. Folder names with slashes are escaped so that each slash in the folder name has a backslash before it. So, to split the path, we need to find slashes with no backslash immediately before them. For this we need a negative lookahead in the regex. javascript doesn't have that. .... but it has negative lookaheads. Soo, we reverse the string and find all slashes that don't have a backslash *after* them. The backslash is now in a negative look-ahead which just makes sure that whatever is at that position is not the thing in the lookahead. After splitting we reverse each part of the path again, in order to be able to read them. Enjoy.

**Comment by marcelklehr** [comment]
Might make sense to put that behavior into a separate function.

**Pull request closed without merging**